infra/vsts_agent_linux_startup.sh

```diff
@@ -95,6 +96,94 @@ useradd \
 #add docker group to user
 usermod -aG docker vsts
 
+# The Bazel cache fills up the hard drive in less than a day, and is a real
+# pain to clean up, as it consists of a very large amount of read-only files in
+# a fairly large tree of read-only directories. Therefore, it looks like the
+# best way to manage it is to mount it on a partition and reset the partition.
+
+cat <<REMOUNT_CACHE > /root/remount_cache.c
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+int exists(const char* path) {
+	struct stat s;
+	return stat(path, &s) == 0 && S_ISDIR(s.st_mode);
+}
+
+int execute_f(char** args) {
+	int status;
+	char** child_env = { NULL };
+	pid_t pid = fork();
+	if (0 == pid) {
+		// child process
+		if (-1 == execve(args[0], args, child_env)) {
+			printf("Failed to start child process calling %s.", args[0]);
+			perror(0);
+			exit(1);
+		}
+		// unreachable: if execve succeeded, we're off to another executable
+	}
+	// parent
+	if (-1 == pid) {
+		printf("Failed to fork.");
+		perror(0);
+		exit(1);
+	}
+	if (-1 == waitpid(pid, &status, 0)) {
+		printf("Failed to wait for child process '%s'.", args[0]);
+		exit(1);
+	}
+	if (1 != WIFEXITED(status)) {
+		printf("Child not terminated: '%s', aborting.", args[0]);
+		exit(1);
+	}
+	if (0 != WEXITSTATUS(status)) {
+		printf("Child '%s' exited with status %d", args[0], WEXITSTATUS(status));
+		exit(WEXITSTATUS(status));
+	}
+	return 0;
+}
+
+#define execute(...) execute_f((char*[]) { __VA_ARGS__, NULL })
+
+int main() {
+	char* mount_point = "/home/vsts/.cache/bazel";
+	char* lost = "/home/vsts/.cache/bazel/lost+found";
+	char* file = "/root/cache_file";
+
+	setuid(0);
+
+	if (exists(lost)) {
+		printf("Removing existing cache...");
+        // This will fail if there are dangling processes, say Bazel servers,
+        // with open fds on the cache.
+		execute("/bin/umount", mount_point);
+	}
+
+	execute("/bin/rm", "-f", file);
```

Comment:
I admit that calling other binaries as root makes me feel very uneasy. Have you considered calling the corresponding syscalls/library functions in process to have a bit more control? I guess you can argue that reimplementing functionality is just as dangerous and those tools are more battleproven but at least you eliminate cases where someone manages to place a binary in `/bin` or `/usr/bin` and while the surface is not quite as battleproven it will be significantly smaller than those binaries.

Reply:
I did consider that, yes, and I did look for those. However, given the interfaces I've seen, it seemed a lot safer (and vastly faster) for me to understand how to safely call a subprocess than for me to understand all the details of each syscall.

As this script does not take any user input, I feel pretty confident about its security. I'm more concerned about it crashing, but even that is not a huge concern as the end result is the same as not having the script (i.e. the machine is unusable and we have to kill it).

I've taken care to not have any user input in calling the child processes (not taking any user input, not going through a shell, not passing through any env var), so AFAIU the only real attack vector here is if someone manages to swap the actual binaries at these hardcoded locations. If someone is already in a position to write as root on the filesystem, they don't really need this script to take over the machine.